### PR TITLE
remove `AMP.extension` calls in production binary

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/5507

### CODE REVIEW

**Comment by erwinmombay** [comment]
friendly ping @dvoytenko (i know you're in jury duty so no rush, just wanted to bring it to your attention in case you didnt see this)

**Comment by jridgewell** [comment]
Why isn't this just using `Amp.push`?

**Comment by erwinmombay** [comment]
@dvoytenko can explain this better than i can

**Comment by dvoytenko** [comment]
@erwinmombay @jridgewell `AMP.extension` is analogous to `AMP.push`. But I don't really understand why they'd be the same? The final interpretation is pretty different: one wraps the whole binary, the other one is only really needed for tests.

File: build-system/runner/src/org/ampproject/AmpPass.java [hunk 1]
```diff
@@ -68,6 +68,8 @@ public AmpPass(AbstractCompiler compiler, boolean isProd,
     if (isCallRemovable(n)) {
       Node rootcall = n.getGrandparent();
       maybeEliminateCallExceptFirstParam(rootcall, rootcall.getParent());
+    } else if (isAddExtensionCall(n)) {
+      inlineAddExtensionCall(n, parent);
     // Remove any `getMode().localDev` and `getMode().test` calls and replace it with `false`.
```
Comment (dvoytenko) [inline build-system/runner/src/org/ampproject/AmpPass.java]: Do we also do `getMode().minimized`?

File: build-system/runner/src/org/ampproject/AmpPass.java [hunk 2]
```diff
@@ -82,6 +84,80 @@ public AmpPass(AbstractCompiler compiler, boolean isProd,
...
+   *     STRING addExtension 3 [length: 12] [source_file: input0]
+   *   STRING some-string 3 [length: 9] [source_file: input0]
+   *   FUNCTION  3 [length: 46] [source_file: input0]
+   */
+  private boolean isAddExtensionCall(Node n) {
+    if (n != null && n.isCall() && n.getChildCount() == 3) {
+      Node getprop = n.getFirstChild();
+
+      // The AST has the last getprop higher in the hierarchy.
+      if (isGetPropName(getprop, "addExtension")) {
```
Comment (dvoytenko) [inline build-system/runner/src/org/ampproject/AmpPass.java]: Why `addExtension`?

File: build-system/runner/test/org/ampproject/AmpPassTest.java [hunk 3]
```diff
@@ -296,4 +296,40 @@ public void testOptimizeGetModeFunction() throws Exception {
              "const IS_SOMETHING = true;",
             "})()"));
   }
+
+  public void testRemoveAmpAddExtensionCallWithExplicitContext() throws Exception {
+    testEs6(
+        LINE_JOINER.join(
+            "var a = 'hello';",
+            "self.AMP.addExtension('hello', function(AMP) {",
```
Comment (dvoytenko) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: Ditto, this should be `AMP.extension`.

File: build-system/runner/test/org/ampproject/AmpPassTest.java [hunk 4]
```diff
@@ -296,4 +296,40 @@ public void testOptimizeGetModeFunction() throws Exception {
...
+            "  var a = 'world';",
+            "  console.log(a);",
+            "});",
+            "console.log(a);"),
+        LINE_JOINER.join(
+            "var a = 'hello';",
+            "(function(AMP) {",
```
Comment (dvoytenko) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: Did we find out if Closure actually unwraps the function completely?

File: build-system/runner/src/org/ampproject/AmpPass.java (hunk 2 above)
Comment (erwinmombay) [inline build-system/runner/src/org/ampproject/AmpPass.java]: my mistake, will fix

File: build-system/runner/test/org/ampproject/AmpPassTest.java (hunk 4 above)
Comment (erwinmombay) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: can you clarify what you mean by unwraps?

Comment (dvoytenko) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: E.g. eventually rewrites this example as:

```
var a = 'hello';
var a2 = 'world';
console.log(a2);
// self.AMP.registerElement(...);
console.log(a);
```

Comment (erwinmombay) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: from the reduction i was testing with, it did not.

Comment (dvoytenko) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: Ok. Makes sense.

@cramforce Not a big deal for you, is it?

File: build-system/runner/src/org/ampproject/AmpPass.java (hunk 2 above)
Comment (erwinmombay) [inline build-system/runner/src/org/ampproject/AmpPass.java]: done.

**Comment by erwinmombay** [comment]
@dvoytenko fixed issues

**Comment by jridgewell** [comment]
My question is why do we need two systems when we already have `Amp.push`? Especially when we have to engineer the second one to DCE.

**Comment by dvoytenko** [comment]
Obviously would be nice to only have one. But they currently work very differently even though the goal is the same. `AMP.push()` wraps the whole binary. `AMP.extension` only wraps side-effects for two reasons:
 1. Remove global side-effects that interfere with tests
 2. Be able to install extensions (side effects) into a desired context.
 3. Still allow module-by-module unit testing.

I'm not yet seeing a way to make `AMP.push` to do both.

**Comment by jridgewell** [comment]
Why don't we get rid of `AMP.push` then? Seems like it's the special case of `AMP.extension`:

```js
global.AMP.extension = function(unusedName, installer) {
  if (IS_DEV) {
    installer(global.AMP);
    return;
  }
  waitForBody(global.document, () => {
    installExtension(fnOrStruct);
  });
};
```

Then we can let normal DCE take care of the dev code.

**Comment by dvoytenko** [comment]
`AMP.push` is hidden from all of the code. It's also used for non-initialized environment, since `[].push` is always available. Likewise, it's good for PROD reasons to wrap the whole extension binary at this time.

**Comment by jridgewell** [comment]
Ok, we can work with `[].push` interface then:

```js
global.AMP.push = function(name, installer) {
  if (IS_DEV) {
    // put this into some object container
    obj[name] = installer;
  }
  waitForBody(global.document, () => {
    installExtension({ n: name, f: installer });
  });
};

for (let i = 0; i < preregisteredExtensions.length; i += 2) {
  const name = preregisteredExtensions[i];
  const installer = preregisteredExtensions[i + 1];
  try {
    installExtension({ n: name, f: installer });
  } catch (e) {
    // Throw errors outside of loop in its own micro task to
    // avoid on error stopping other extensions from loading.
    dev().error(TAG, 'Extension failed: ', e, name);
  }
}

File: build-system/runner/test/org/ampproject/AmpPassTest.java (hunk 4 above)
Comment (erwinmombay) [inline build-system/runner/test/org/ampproject/AmpPassTest.java]: @cramforce it doesn't with what we currently have

**Comment by erwinmombay** [comment]
ping @dvoytenko for response

**Comment by dvoytenko** [comment]
@erwinmombay @jridgewell I still don't quiet see how we'd use `AMP.push` here. How would we know which `AMP.push` to keep and which one can be DCE'd? `AMP.push` is still a valid pattern in the code, though rare.

**Comment by jridgewell** [comment]
There's no reason to DCE `AMP.push` calls, but its code can be DCE because `IS_DEV` will statically be `false` in prod builds.

**Comment by dvoytenko** [comment]
Hmm. In that case, it'd be better to do a different pattern:

```
function install(AMP) {
  ...
}

if (IS_DEV) {
  AMP.push({n: 'ext1', f: install});
} else {
  install(AMP);
}
```

A little more verbose. But then we have everything ready for it already.
WDYT?

**Comment by jridgewell** [comment]
Really, I want to wrap all our code in `AMP.push` so we get rid of the build step changes. 😁

**Comment by erwinmombay** [comment]
yeah i wanted to avoid having to do this on the visible code side and on the closure end too and just wrap the whole thing up, but its hard because of the usage where we directly import these modules in test code.
 
let me look at it again if we can do that wrapping on the babel/browserify end. Also IS_DEV only currently works if its in a function and doesn't work on the module level (easy enough to get to work, the problem is that on the module level it is actually changed to something like $path$to$module$IS_DEV to avoid collisions when all the modules are hoisted to a single scope for tree shaking)

**Comment by cramforce** [comment]
Wrapping in user land would only work if all deps are 100% side effect free. I'd prefer a wrapper around all code for various reasons.

**Pull request merged**